src/common/cups_print.c

```diff
@@ -323,61 +397,136 @@ void dt_print_file(const int32_t imgid, const char *filename, const dt_print_inf
     return;
   }
 
-  cups_dest_t *dests;
-  int num_dests = cupsGetDests(&dests);
-  cups_dest_t *dest = cupsGetDest(pinfo->printer.name, NULL, num_dests, dests);
-
   cups_option_t *options = NULL;
   int num_options = 0;
 
-  for (int j = 0; j < dest->num_options; j ++)
-    if (cupsGetOption(dest->options[j].name, num_options,
-                      options) == NULL)
-      num_options = cupsAddOption(dest->options[j].name,
-                                  dest->options[j].value,
-                                  num_options, &options);
+  // for turboprint drived printer, use the turboprint dialog
+  if (pinfo->printer.is_turboprint)
+  {
+    const char *tp_intent_name[] = { "perception_0", "colorimetric-relative_1", "saturation_1", "colorimetric-absolute_1" };
+    char tpcmd[1024];
+    char tmpfile[PATH_MAX] = { 0 };
 
-  cupsFreeDests(num_dests, dests);
+    dt_loc_get_tmp_dir(tmpfile, sizeof(tmpfile));
+    g_strlcat(tmpfile, "/dt_cups_opts_XXXXXX", sizeof(tmpfile));
 
-  // if we have a profile, disable cm on CUPS, this is important as dt does the cm
+    gint fd = g_mkstemp(tmpfile);
+    if(fd == -1)
+    {
+      dt_control_log(_("failed to create temporary file for printing options"));
+      fprintf(stderr, "failed to create temporary pdf for printing options\n");
+      return;
+    }
+    close(fd);
 
-  num_options = cupsAddOption("cm-calibration", *pinfo->printer.profile ? "true" : "false", num_options, &options);
+    // ensure that intent is in the range, may happen if at some point we add new intent in the list
+    const int intent = (pinfo->printer.intent < 4) ? pinfo->printer.intent : 0;
 
-  // media to print on
+    // start the turboprint dialog
+    if (snprintf
+        (tpcmd, sizeof(tpcmd),
+         "turboprint --printer=%s --options --output=%s -o copies=1 -o PageSize=%s -o InputSlot=AutoSelect -o zedoIntent=%s -o MediaType=%s",
+         pinfo->printer.name, tmpfile, pinfo->paper.common_name, tp_intent_name[intent], pinfo->medium.name)
```

Comment:
Note how you are printing `tmpfile` (which is `PATH_MAX`, i.e. `4096`) into `tpcmd`, which is just `1024`.
It might be better to use `g_strdup_printf()` or something from the start.

Reply:
Indeed, lot better.